lib/portlayer/event/manager.go

```diff
@@ -0,0 +1,158 @@
+	"github.com/vmware/vic/lib/portlayer/event/events"
+)
+
+type Manager struct {
+	cos  collectorCache
+	subs subscriberCache
+}
+
+type collectorCache struct {
+	mu sync.RWMutex
+
+	collectors map[string]collector.Collector
+}
+
+type subscriberCache struct {
+	mu sync.RWMutex
+
+	subscribers map[interface{}]map[string]func(events.Event)
+}
+
+func NewEventManager(collectors ...collector.Collector) *Manager {
+	mgr := &Manager{
+		cos: collectorCache{
+			collectors: make(map[string]collector.Collector),
+		},
+		subs: subscriberCache{
+			subscribers: make(map[interface{}]map[string]func(events.Event)),
+		},
+	}
+
+	// register any collectors provided
+	for i := range collectors {
+		mgr.RegisterCollector(collectors[i])
+	}
+
+	return mgr
+}
+
+func (mgr *Manager) RegisterCollector(collector collector.Collector) {
+	if collector == nil {
+		return
+	}
+
+	mgr.cos.mu.Lock()
+	defer mgr.cos.mu.Unlock()
+
+	collector.Register(mgr.Publish)
+
+	mgr.cos.collectors[collector.Name()] = collector
+}
+
+func (mgr *Manager) Collectors() map[string]collector.Collector {
+	mgr.cos.mu.RLock()
+	defer mgr.cos.mu.RUnlock()
+
+	c := make(map[string]collector.Collector)
+	for name, collector := range mgr.cos.collectors {
+		c[name] = collector
+	}
+	return c
+}
+
+// Subscribe to the event manager for callback
+func (mgr *Manager) Subscribe(eventTopic interface{}, caller string, callback func(events.Event)) {
+	mgr.subs.mu.Lock()
+	defer mgr.subs.mu.Unlock()
+	t := topic(eventTopic)
+	if _, ok := mgr.subs.subscribers[t]; !ok {
+
+		mgr.subs.subscribers[t] = make(map[string]func(events.Event))
+	}
+	mgr.subs.subscribers[t][caller] = callback
+}
+
+// Unsubscribe from callbacks
+func (mgr *Manager) Unsubscribe(eventTopic interface{}, caller string) {
+	mgr.subs.mu.Lock()
+	defer mgr.subs.mu.Unlock()
+	t := topic(eventTopic)
+	if _, ok := mgr.subs.subscribers[t]; ok {
+		delete(mgr.subs.subscribers[t], caller)
+	}
+}
+
+func (mgr *Manager) Subscribers() map[interface{}]map[string]func(events.Event) {
+	mgr.subs.mu.RLock()
+	defer mgr.subs.mu.RUnlock()
+	s := make(map[interface{}]map[string]func(events.Event))
+	for i, m := range mgr.subs.subscribers {
+
+		if _, ok := s[i]; !ok {
+			s[i] = make(map[string]func(events.Event))
+		}
+
+		for k, v := range m {
+			s[i][k] = v
+		}
+	}
+	return s
+}
+
+// RegistryCount returns the callback count
+func (mgr *Manager) Subscribed() int {
+	mgr.subs.mu.RLock()
+	defer mgr.subs.mu.RUnlock()
+	count := 0
+	for _, m := range mgr.subs.subscribers {
+		count += len(m)
+	}
+	return count
+}
+
+// Publish events to subscribers
+func (mgr *Manager) Publish(e events.Event) {
+
+	// default value of interface
+	t := topic(e)
+
+	// subscribers for this event
+	subs := mgr.subs.subscribers[t]
+
+	log.Debugf("Found %d subscribers to %s", len(subs), reflect.TypeOf(t))
+
+	for sub, f := range subs {
+		//TODO: need to introduce timeout -- either with callbacks
+		// or when channels are introduced
```

Comment:
Open an issue. I think we'll settle on goroutine given how simple it is to fire. If we want to be really proper we can pass a clone of the event to avoid concurrent modification issues.